File: bridge_app/lib/services/clipboard_history_service.dart
import 'dart:convert';
import 'dart:io';
import 'package:flutter/foundation.dart';
import 'package:flutter/services.dart';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'package:uuid/uuid.dart';
import 'system_channel.dart';
import 'clipboard_service.dart';
import 'content_classifier.dart';

const int kMaxClipboardHistory = 20;
const String _kStorageKey = 'clipboard_history';

class ClipboardHistoryItem {
  final String id;
  final String kind; // 'text' | 'image'
  final String contentType; // 'url' | 'otp' | 'email' | 'phone' | 'text' | 'image'
  final String? text;
  final String? imageThumbnail; // Base64 thumbnail or null
  final String? imagePath;      // Absolute local cache path
  final DateTime timestamp;
  final String origin; // 'android' | 'windows'

  const ClipboardHistoryItem({
    required this.id,
    this.kind = 'text',
    this.contentType = 'text',
    this.text,
    this.imageThumbnail,
    this.imagePath,
    required this.timestamp,
    required this.origin,
  });

  Map<String, dynamic> toJson() => {
        'id': id,
        'kind': kind,
        'contentType': contentType,
        'text': text,
        'imageThumbnail': imageThumbnail,
        'imagePath': imagePath,
        'timestamp': timestamp.toUtc().toIso8601String(),
        'origin': origin,
      };

  factory ClipboardHistoryItem.fromJson(Map<String, dynamic> json) {
    final kind = json['kind'] as String? ?? 'text';
    final rawText = json['text'] as String?;
    final contentType = json['contentType'] as String? ??
        (kind == 'image' ? 'image' : classifyClipboardText(rawText ?? ''));

    return ClipboardHistoryItem(
      id: json['id'] as String? ?? const Uuid().v4(),
      kind: kind,
      contentType: contentType,
      text: rawText,
      imageThumbnail: json['imageThumbnail'] as String?,
      imagePath: json['imagePath'] as String?,
      timestamp: json['timestamp'] != null
          ? DateTime.tryParse(json['timestamp'] as String)?.toLocal() ?? DateTime.now()
          : DateTime.now(),
      origin: json['origin'] as String? ?? 'android',
    );
  }
}

class ClipboardHistoryService {
  ClipboardHistoryService._();
  static final ClipboardHistoryService instance = ClipboardHistoryService._();

  final _storage = const FlutterSecureStorage();
  final ValueNotifier<List<ClipboardHistoryItem>> items = ValueNotifier([]);
  bool _initialized = false;

  Future<void> init() async {
    if (_initialized) return;
    _initialized = true;
    await load();
    await _pruneOrphanedImages();
  }

  Future<void> load() async {
    try {
      final raw = await _storage.read(key: _kStorageKey);
      if (raw != null && raw.isNotEmpty) {
        final decoded = jsonDecode(raw) as List<dynamic>;
        final parsed = decoded
            .map((e) => ClipboardHistoryItem.fromJson(Map<String, dynamic>.from(e as Map)))
            .take(kMaxClipboardHistory)
            .toList();
        items.value = List.unmodifiable(parsed);
        debugPrint('[ClipboardHistory] Loaded ${parsed.length} items from storage');
      }
    } catch (e) {
      debugPrint('[ClipboardHistory] Error loading history: $e');
    }
  }

  Future<void> addEntry(
    String text,
    String origin, {
    String? id,
    DateTime? timestamp,
    String? contentType,
  }) async {
    if (text.trim().isEmpty) return;

    final currentList = List<ClipboardHistoryItem>.from(items.value);

    // Suppress consecutive identical text item at the top
    if (currentList.isNotEmpty &&
        currentList.first.kind == 'text' &&
        currentList.first.text == text &&
        currentList.first.origin == origin) {
      return;
    }

    final resolvedContentType = contentType ?? classifyClipboardText(text);

    final newItem = ClipboardHistoryItem(
      id: id ?? const Uuid().v4(),
      kind: 'text',
      contentType: resolvedContentType,
      text: text,
      timestamp: timestamp ?? DateTime.now(),
      origin: origin,
    );

    _insertAndTrim(newItem, currentList);
  }

  Future<void> addImageEntry({
    required String imagePath,
    required String origin,
    String? id,
    String? imageThumbnail,
    DateTime? timestamp,
  }) async {
    final currentList = List<ClipboardHistoryItem>.from(items.value);

    // Suppress consecutive identical image item at the top
    if (currentList.isNotEmpty &&
        currentList.first.kind == 'image' &&
        currentList.first.imagePath == imagePath &&
        currentList.first.origin == origin) {
      return;
    }

    final newItem = ClipboardHistoryItem(
      id: id ?? const Uuid().v4(),
      kind: 'image',
      contentType: 'image',
      imagePath: imagePath,
      imageThumbnail: imageThumbnail,
      timestamp: timestamp ?? DateTime.now(),
      origin: origin,
    );

    _insertAndTrim(newItem, currentList);
  }

  void _insertAndTrim(ClipboardHistoryItem newItem, List<ClipboardHistoryItem> currentList) async {
    // Prepend to list without reordering or removing existing items
    currentList.insert(0, newItem);

    // Cap at 20 entries (oldest drops off)
    if (currentList.length > kMaxClipboardHistory) {
      final evicted = currentList.sublist(kMaxClipboardHistory);
      for (final ev in evicted) {
        _deleteImageFile(ev.imagePath);
      }
    }

    final capped = currentList.take(kMaxClipboardHistory).toList();
    items.value = List.unmodifiable(capped);

    await _save(capped);
    debugPrint('[ClipboardHistory] Added new ${newItem.kind} entry (${newItem.origin}): ${newItem.contentType} (${capped.length}/20)');
  }

  /// Copies an item back to Android's local clipboard without re-syncing to Windows.
  Future<void> copyLocally(ClipboardHistoryItem item) async {
    if (item.kind == 'image' && item.imagePath != null) {
      ClipboardService.instance.markImageAsSynced(item.imagePath!);
      await SystemChannel.setClipboardImage(item.imagePath!);
      debugPrint('[ClipboardHistory] Copied image history item locally: ${item.imagePath} (sync suppressed)');
      return;
    }

    final text = item.text ?? '';
    if (text.isEmpty) return;

    // 1. Tell ClipboardService to track this as last-synced so resume sync doesn't re-emit
    ClipboardService.instance.markAsSynced(text);

    // 2. Set native clipboard directly
    await SystemChannel.setClipboard(text);

    // 3. Also set Flutter platform clipboard for current Activity
    try {
      await Clipboard.setData(ClipboardData(text: text));
    } catch (e) {
      debugPrint('[ClipboardHistory] Clipboard.setData error: $e');
    }

    debugPrint('[ClipboardHistory] Copied history item locally: "${text.length > 40 ? '${text.substring(0, 40)}…' : text}" (sync suppressed)');
  }

  /// Backward-compatible string copy helper
  Future<void> copyLocallyText(String text) async {
    if (text.isEmpty) return;
    ClipboardService.instance.markAsSynced(text);
    await SystemChannel.setClipboard(text);
    try {
      await Clipboard.setData(ClipboardData(text: text));
    } catch (e) {
      debugPrint('[ClipboardHistory] Clipboard.setData error: $e');
    }
  }

  Future<void> clear() async {
    for (final item in items.value) {
      _deleteImageFile(item.imagePath);
    }
    items.value = const [];
    await _storage.delete(key: _kStorageKey);
  }

  void _deleteImageFile(String? path) {
    if (path == null || path.isEmpty) return;
    try {
      final f = File(path);
      if (f.existsSync()) {
        f.deleteSync();
        debugPrint('[ClipboardHistory] Evicted & deleted image file: $path');
      }
    } catch (e) {
      debugPrint('[ClipboardHistory] Failed to delete image file $path: $e');
    }
  }

  Future<void> _pruneOrphanedImages() async {
    try {
      final cacheDir = await SystemChannel.getClipboardCacheDir();
      if (cacheDir == null) return;
      final dir = Directory(cacheDir);
      if (!dir.existsSync()) return;

      final activePaths = items.value
          .where((i) => i.kind == 'image' && i.imagePath != null)
          .map((i) => File(i.imagePath!).absolute.path)
          .toSet();

      final entities = dir.listSync();
      for (final entity in entities) {
        if (entity is File && !activePaths.contains(entity.absolute.path)) {
          try {
            entity.deleteSync();
            debugPrint('[ClipboardHistory] Pruned orphaned cached image: ${entity.path}');
          } catch (_) {}
        }
      }
    } catch (e) {
      debugPrint('[ClipboardHistory] Error during orphaned image prune: $e');
    }
  }

  Future<void> _save(List<ClipboardHistoryItem> list) async {
    try {
      final jsonList = list.map((e) => e.toJson()).toList();
      await _storage.write(key: _kStorageKey, value: jsonEncode(jsonList));
    } catch (e) {
      debugPrint('[ClipboardHistory] Error saving history: $e');
    }
  }
}

File: bridge-agent/electron/services/ClipboardHistoryService.ts
import { app } from 'electron'
import fs from 'node:fs'
import path from 'node:path'
import { randomUUID } from 'node:crypto'
import type { ClipboardHistoryItem, ClipboardContentType, Origin } from '../types/protocol.js'
import { classifyClipboardText } from './contentClassification.js'

const MAX_HISTORY_ITEMS = 20

class ClipboardHistoryServiceClass {
  private _items: ClipboardHistoryItem[] = []
  private _filePath: string = ''
  private _imagesDir: string = ''
  private _listeners: Array<(items: ClipboardHistoryItem[]) => void> = []

  init(): void {
    try {
      const userData = app.getPath('userData')
      this._filePath = path.join(userData, 'clipboard-history.json')
      this._imagesDir = path.join(userData, 'clipboard-images')
      fs.mkdirSync(this._imagesDir, { recursive: true })
      this._load()
      this._pruneOrphanedImages()
      console.log(`[ClipboardHistory] Initialized with ${this._items.length} items from ${this._filePath}`)
    } catch (err) {
      console.warn('[ClipboardHistory] Failed to initialize history storage:', err)
    }
  }

  getImagesDir(): string {
    return this._imagesDir
  }

  getItems(): ClipboardHistoryItem[] {
    return [...this._items]
  }

  getItemById(id: string): ClipboardHistoryItem | undefined {
    return this._items.find((i) => i.id === id)
  }

  addEntry(
    text: string,
    origin: Origin,
    timestamp?: string,
    id?: string,
    contentType?: ClipboardContentType
  ): ClipboardHistoryItem | null {
    if (!text || text.trim() === '') return null

    // If top item has exact same text and origin, don't duplicate
    if (this._items.length > 0 && this._items[0].kind === 'text' && this._items[0].text === text && this._items[0].origin === origin) {
      return null
    }

    const item: ClipboardHistoryItem = {
      id: id || randomUUID(),
      kind: 'text',
      contentType: contentType || classifyClipboardText(text),
      text,
      timestamp: timestamp || new Date().toISOString(),
      origin,
    }

    this._insertAndTrim(item)
    return item
  }

  addImageEntry(
    imageThumbnail: string,
    imagePath: string,
    origin: Origin,
    timestamp?: string,
    id?: string
  ): ClipboardHistoryItem | null {
    // If top item has exact same image path and origin, don't duplicate
    if (this._items.length > 0 && this._items[0].kind === 'image' && this._items[0].imagePath === imagePath && this._items[0].origin === origin) {
      return null
    }

    const item: ClipboardHistoryItem = {
      id: id || randomUUID(),
      kind: 'image',
      contentType: 'image',
      imageThumbnail,
      imagePath,
      timestamp: timestamp || new Date().toISOString(),
      origin,
    }

    this._insertAndTrim(item)
    return item
  }

  private _insertAndTrim(item: ClipboardHistoryItem): void {
    // Prepend to list without reordering or removing other existing items
    this._items.unshift(item)

    // Cap at MAX_HISTORY_ITEMS (oldest drops off)
    if (this._items.length > MAX_HISTORY_ITEMS) {
      const evicted = this._items.slice(MAX_HISTORY_ITEMS)
      this._items = this._items.slice(0, MAX_HISTORY_ITEMS)
      for (const ev of evicted) {
        this._deleteImageFile(ev.imagePath)
      }
    }

    this._save()
    this._notifyListeners()
  }

  onUpdate(callback: (items: ClipboardHistoryItem[]) => void): void {
    this._listeners.push(callback)
  }

  clear(): void {
    for (const item of this._items) {
      this._deleteImageFile(item.imagePath)
    }
    this._items = []
    this._save()
    this._notifyListeners()
  }

  private _deleteImageFile(imagePath?: string): void {
    if (!imagePath) return
    try {
      if (fs.existsSync(imagePath)) {
        fs.unlinkSync(imagePath)
        console.log(`[ClipboardHistory] Evicted & unlinked image file: ${imagePath}`)
      }
    } catch (e) {
      console.warn(`[ClipboardHistory] Failed to unlink image file ${imagePath}:`, e)
    }
  }

  private _pruneOrphanedImages(): void {
    if (!this._imagesDir || !fs.existsSync(this._imagesDir)) return
    try {
      const activePaths = new Set(
        this._items
          .filter((i) => i.kind === 'image' && !!i.imagePath)
          .map((i) => path.resolve(i.imagePath!))
      )
      const files = fs.readdirSync(this._imagesDir)
      for (const file of files) {
        const fullPath = path.resolve(path.join(this._imagesDir, file))
        if (!activePaths.has(fullPath)) {
          try {
            fs.unlinkSync(fullPath)
            console.log(`[ClipboardHistory] Pruned orphaned cached image: ${fullPath}`)
          } catch {}
        }
      }
    } catch (e) {
      console.warn('[ClipboardHistory] Error during orphaned image prune:', e)
    }
  }

  private _notifyListeners(): void {
    for (const listener of this._listeners) {
      try {
        listener(this.getItems())
      } catch (e) {
        console.error('[ClipboardHistory] Error notifying listener:', e)
      }
    }
  }

  private _load(): void {
    if (!this._filePath || !fs.existsSync(this._filePath)) {
      this._items = []
      return
    }

    try {
      const content = fs.readFileSync(this._filePath, 'utf8')
      const parsed = JSON.parse(content)
      if (Array.isArray(parsed)) {
        this._items = parsed.slice(0, MAX_HISTORY_ITEMS).map((raw) => {
          const item = raw as Partial<ClipboardHistoryItem>
          return {
            id: item.id || randomUUID(),
            kind: item.kind || 'text',
            contentType: item.contentType || (item.kind === 'image' ? 'image' : classifyClipboardText(item.text || '')),
            text: item.text,
            imageThumbnail: item.imageThumbnail,
            imagePath: item.imagePath,
            timestamp: item.timestamp || new Date().toISOString(),
            origin: item.origin || 'windows',
          }
        })
      } else {
        this._items = []
      }
    } catch (err) {
      console.warn('[ClipboardHistory] Failed to parse history file, starting fresh:', err)
      this._items = []
    }
  }

  private _save(): void {
    if (!this._filePath) return
    try {
      fs.writeFileSync(this._filePath, JSON.stringify(this._items, null, 2), 'utf8')
    } catch (err) {
      console.error('[ClipboardHistory] Failed to save history to file:', err)
    }
  }
}

export const ClipboardHistoryService = new ClipboardHistoryServiceClass()
